pkg/agent/qrm-plugins/cpu/dynamicpolicy/mbm/control.go

```diff
@@ -0,0 +1,235 @@
+/*
+Copyright 2022 The Katalyst Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mbm
+
+import (
+	"context"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/kubewharf/katalyst-core/pkg/agent/qrm-plugins/cpu/dynamicpolicy/mbm/strategy"
+	"github.com/kubewharf/katalyst-core/pkg/agent/qrm-plugins/cpu/dynamicpolicy/state"
+	"github.com/kubewharf/katalyst-core/pkg/metaserver/agent/metric/types"
+	"github.com/kubewharf/katalyst-core/pkg/metrics"
+	"github.com/kubewharf/katalyst-core/pkg/util/external/mbm"
+	"github.com/kubewharf/katalyst-core/pkg/util/general"
+)
+
+const (
+	MemoryBandwidthManagement = "mbm"
+
+	// controlling hype-parameters in percentage
+	MEMORY_BANDWIDTH_PHYSICAL_NUMA_PAINPOINT       = 105
+	MEMORY_BANDWIDTH_PHYSICAL_NUMA_SWEETPOINT      = 95
+	MEMORY_BANDWIDTH_PHYSICAL_NUMA_UNTHROTTLEPOINT = 80
+
+	MEMORY_BANDWIDTH_INCREASE_MIN = 256 // the minimum incremental value
+	MEMORY_BANDWIDTH_DECREASE_MIN = 512 // the minimum decremental value
+)
+
+type NUMAStater interface {
+	GetMachineState() state.NUMANodeMap
+}
+
+type Controller struct {
+	metricEmitter metrics.MetricEmitter
+	metricReader  types.MetricsReader
+	numaStater    NUMAStater
+	mbAdjust      mbm.MBAdjuster
+
+	numaThrottled sets.Int
+
+	packageMap         map[int][]int // package id --> numa nodes in the package
+	interval           time.Duration
+	bandwidthThreshold int64
+	minDeductionStep   float64
+	minIncreaseStep    float64
+}
+
+func (c Controller) Run(ctx context.Context) {
+	general.Infof("mbm controller is starting")
+	wait.Until(c.run, c.interval, ctx.Done())
+}
+
+func (c Controller) run() {
+	for p, nodes := range c.packageMap {
+		c.processPackage(p, nodes)
+	}
+}
+
+// getActiveNodeSets divides the nodes into workload-associated sets
+// only nodes having active workload assigned shall be considered
+func (c Controller) getActiveNodeSets(nodes []int) map[string]sets.Int {
```

Comment:
filter dedicated_cores with numa_binding pods

Reply:
for now only socket type pods are assumed on the machine - which is a limitation. Will add support of mixed pod types in the subsequent change AFTER the current code be validated. I will add comments to code to make this assumption and plan clear. Let me know if code comment not good enough in this case.